### PR TITLE
Rewriting postorder and preorder

### PR DESCRIPTION
**Pull request by Pihjoe**
### Summary

Refactored postorder and preorder to be more like client notes, plus tests and necessary changes to other functions that uses those.

### CODE REVIEW

File: test/python/transpiler/test_graph_utils.py [hunk 1]
```diff
@@ -71,20 +70,42 @@ def test_postorder_traversal_returns_empty_list_for_tree_with_no_edges(self):
...
         tree.add_nodes_from([0])
-        result = []
-        postorder_traversal(tree, 0, result)
+
+        result = postorder_traversal(tree, 0)
 
         self.assertEqual(result, [])
 
+    def test_postorder_traversal_returns_correct_edges_if_visited_list_is_not_empty(self):
+        """Test that postorder_traversal returns correct edge list when some nodes have laready been visited"""
```
Comment (seirepo) [inline test/python/transpiler/test_graph_utils.py:79]: There's a typo, but otherwise this pull request looks pretty good!

**Comment by seirepo** [comment]
A small note: maybe there could be a test for the case when the tree is not actually a tree but has a loop, just to see that nothing explodes if that's the case? I guess it shouldn't happen with this code

**Comment by MillaKelhu** [comment]
> A small note: maybe there could be a test for the case when the tree is not actually a tree but has a loop, just to see that nothing explodes if that's the case? I guess it shouldn't happen with this code

`postorder_traversal()` and `preorder_traversal()` are only called with the input tree being constructed with `rx.steiner_tree()`, and a steiner tree should never have any loops, so I think that the test would be unnecessary.

**Comment by seirepo** [comment]
That's true, I was wondering if someone else uses them with their own input graph but maybe it's the caller's responsibility

**Review by seirepo — APPROVED** [review]

**Review by Alex-Elias — APPROVED** [review]

**Pull request merged**